pkg/controllers/provisioning/resourcecounter.go

```diff
@@ -0,0 +1,85 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"context"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha5"
+	"gopkg.in/inf.v0"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ResourceCount struct {
+	CPU    *inf.Dec
+	Memory *inf.Dec
+}
+
+type ResourceCounter struct {
+	KubeClient client.Client
+}
+
+// Constructs a new instance of the resource counter
+func NewResourceCounter(kubeClient client.Client) *ResourceCounter {
+	return &ResourceCounter{
+		KubeClient: kubeClient,
+	}
+}
+
+func (r *ResourceCounter) remainingResourceCounts(ctx context.Context, limits v1alpha5.Limits, provisionerName string) (*ResourceCount, error) {
+	// We recalculate remaining resources each provisioning loop, because the terminate controller could've freed up some capacity since
+	// we last provisioned worker nodes.
+	nodeList := v1.NodeList{}
+	withProvisionerName := client.MatchingLabels{v1alpha5.ProvisionerNameLabelKey: provisionerName}
+	if err := r.KubeClient.List(ctx, &nodeList, withProvisionerName); err != nil {
+		return nil, err
+	}
+	resourceCount := ResourceCount{
+		CPU:    resource.Zero.AsDec(),
+		Memory: resource.Zero.AsDec(),
+	}
+	for _, node := range nodeList.Items {
+		resourceCount.CPU.Add(resourceCount.CPU, node.Status.Capacity.Cpu().AsDec())
+		resourceCount.Memory.Add(resourceCount.Memory, node.Status.Capacity.Memory().AsDec())
+	}
```

Comment:
Custom resources like GPUs aren't going to show up in the nodeSpec. So it's difficult to determine how many GPUs have been provisioned by the cluster like this. 

I'm likely going to do a translation that goes nodeObject -> instanceTypeLabel -> retrieveInstanceTypeInfo. Figuring out this bit.